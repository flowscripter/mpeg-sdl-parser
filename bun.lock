{
  "lockfileVersion": 1,
  "workspaces": {
    "": {
      "name": "@flowscripter/mpeg-sdl-parser",
      "dependencies": {
        "typescript-parsec": "latest",
      },
      "devDependencies": {
        "@types/bun": "latest",
      },
      "peerDependencies": {
        "typescript": "latest",
      },
    },
  },
  "packages": {
    "@types/bun": ["@types/bun@1.2.10", "", { "dependencies": { "bun-types": "1.2.10" } }, "sha512-eilv6WFM3M0c9ztJt7/g80BDusK98z/FrFwseZgT4bXCq2vPhXD4z8R3oddmAn+R/Nmz9vBn4kweJKmGTZj+lg=="],

    "@types/node": ["@types/node@22.13.10", "", { "dependencies": { "undici-types": "~6.20.0" } }, "sha512-I6LPUvlRH+O6VRUqYOcMudhaIdUVWfsjnZavnsraHvpBwaEyMN29ry+0UVJhImYL16xsscu0aske3yA+uPOWfw=="],

    "bun-types": ["bun-types@1.2.10", "", { "dependencies": { "@types/node": "*" } }, "sha512-b5ITZMnVdf3m1gMvJHG+gIfeJHiQPJak0f7925Hxu6ZN5VKA8AGy4GZ4lM+Xkn6jtWxg5S3ldWvfmXdvnkp3GQ=="],

    "typescript": ["typescript@5.8.3", "", { "bin": { "tsc": "bin/tsc", "tsserver": "bin/tsserver" } }, "sha512-p1diW6TqL9L07nNxvRMM7hMMw4c5XOo/1ibL4aAIGmSAt9slTE1Xgw5KWuof2uTOvCg9BY7ZRi+GaF+7sfgPeQ=="],

    "typescript-parsec": ["typescript-parsec@0.3.4", "", {}, "sha512-6RD4xOxp26BTZLopNbqT2iErqNhQZZWb5m5F07/UwGhldGvOAKOl41pZ3fxsFp04bNL+PbgMjNfb6IvJAC/uYQ=="],

    "undici-types": ["undici-types@6.20.0", "", {}, "sha512-Ny6QZ2Nju20vw1SRHe3d9jVu6gJ+4e3+MMpqu7pqE5HT6WsTSlce++GQmK5UXS8mzV8DSYHrQH+Xrf2jVcuKNg=="],
  }
}
