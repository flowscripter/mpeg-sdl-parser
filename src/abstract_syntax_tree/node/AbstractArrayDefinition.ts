import type Location from "../../tokenizer/token/Location.ts";
import type SyntaxToken from "../../tokenizer/token/SyntaxToken.ts";
import AbstractStatement from "./AbstractStatement.ts";
import type StatementKind from "./enum/statement_kind.ts";
import type Identifier from "./Identifier.ts";

abstract class AbstractArrayDefinition extends AbstractStatement {
  constructor(
    kind: StatementKind,
    location: Location,
    public readonly identifier: Identifier,
    public readonly semicolonPunctuatorToken: SyntaxToken,
  ) {
    super(kind, location);
  }
}

export default AbstractArrayDefinition;
