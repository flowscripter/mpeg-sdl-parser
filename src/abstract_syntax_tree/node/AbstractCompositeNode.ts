import type Location from "../../tokenizer/token/Location.ts";
import type NodeVisitor from "../visitor/NodeVisitor.ts";
import AbstractNode from "./AbstractNode.ts";
import type NodeKind from "./enum/node_kind.ts";

abstract class AbstractCompositeNode extends AbstractNode {
  constructor(
    nodeKind: NodeKind,
    location: Location,
  ) {
    super(nodeKind, location, true);
  }

  override accept(visitor: NodeVisitor): boolean {
    if (!visitor.visitBefore(this)) {
      return false;
    }

    for (const child of this.getChildNodeIterable()) {
      if (!child.accept(visitor)) {
        break;
      }
    }

    return visitor.visitAfter(this);
  }

  abstract getChildNodeIterable(): IterableIterator<AbstractNode>;
}

export default AbstractCompositeNode;
