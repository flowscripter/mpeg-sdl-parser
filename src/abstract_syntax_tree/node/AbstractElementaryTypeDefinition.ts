import type { Location } from "../../tokenizer/token/Location.ts";
import type { SyntaxToken } from "../../tokenizer/token/SyntaxToken.ts";
import type { AbstractNode } from "./AbstractNode.ts";
import { AbstractStatement } from "./AbstractStatement.ts";
import type { ElementaryType } from "./ElementaryType.ts";
import type { StatementKind } from "./enum/statement_kind.ts";
import type { Identifier } from "./Identifier.ts";

export abstract class AbstractElementaryTypeDefinition
  extends AbstractStatement {
  constructor(
    kind: StatementKind,
    location: Location,
    public readonly isConst: boolean,
    public readonly elementaryType: ElementaryType,
    public readonly identifier: Identifier,
    public readonly value: AbstractNode | undefined,
    public readonly constToken: SyntaxToken | undefined,
    public readonly assignmentToken: SyntaxToken | undefined,
    public readonly semicolonPunctuatorToken: SyntaxToken,
  ) {
    super(kind, location);
  }
}
