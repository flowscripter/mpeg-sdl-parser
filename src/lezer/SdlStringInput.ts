import type { Input } from "@lezer/common";

/**
 * The `SdlStringInput` class implements the `Input` interface for a string input.
 * It provides methods to read and chunk the input string.
 */
export default class SdlStringInput implements Input {
  private readonly input: string;

  constructor(input: string) {
    this.input = input.replace(/\r\n?/g, "\n");
  }

  /**
    The length of the document.
    */
  get length() {
    return this.input.length;
  }

  /**
   * Get the chunk from the given position. The returned string
   * starts at `from` (0-based) and, if that isn't the end of the
   * string, may be of any length greater than zero.
   *
   * @param from - The starting position of the chunk (0-based, inclusive).
   */
  chunk(from: number): string {
    return this.input.slice(from);
  }

  /**
   * Indicates that the chunks do not already end at line breaks and that
   * client code that wants to work by-line must scanning for line breaks.
   */
  lineChunks = false;

  /**
   * Read the part of the document between the given positions.
   * @param from - The starting position (0-based, inclusive).
   * @param to - The ending position (0-based, exclusive).
   */
  read(from: number, to: number): string {
    return this.input.slice(from, to);
  }
}
