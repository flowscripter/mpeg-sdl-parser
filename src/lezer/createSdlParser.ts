import { buildParser } from "@lezer/generator";
import { LRParser as LezerParser } from "@lezer/lr";
import fs from "node:fs/promises";
import path from "node:path";
import getLogger from "../util/logger.ts";
import { createSyntacticParseError } from "../ParseError.ts";
import { Text } from "@codemirror/state";
import type { Input } from "@lezer/common";

const sdlParserLogger = getLogger("SdlParser");

let lenientSdlParser: LezerParser | undefined;
let strictSdlParser: LezerParser | undefined;

/**
 * Create an in memory lenient Lezer based parser using the SDL grammar and store it as a "singleton".
 */
export async function createLenientSdlParser(): Promise<LezerParser> {
  if (!lenientSdlParser) {
    const fileName = "sdl.lezer.grammar";
    const grammarPath = new URL(
      path.join("../..", "grammar", fileName),
      import.meta.url,
    );
    const grammarText = await fs.readFile(grammarPath, "utf-8");

    lenientSdlParser = buildParser(grammarText, {
      fileName,
      warn: sdlParserLogger.warn,
    });
  }

  return lenientSdlParser;
}

/**
 * Create an in memory strict Lezer based parser using the SDL grammar and store it as a "singleton".
 */
export async function createStrictSdlParser(): Promise<LezerParser> {
  if (!strictSdlParser) {
    strictSdlParser = (await createLenientSdlParser()).configure({
      strict: true,
    });

    // Wrap the parser's parse method to catch errors and wrap them in ParseError
    const originalParseFunc = strictSdlParser.parse.bind(strictSdlParser);

    strictSdlParser.parse = (input: Input | string, ...args) => {
      try {
        return originalParseFunc(input, ...args);
      } catch (err) {
        // parse position out of error message which takes the form "No parse at 0"
        const errorMessage = (err as Error).message;
        const match = errorMessage.match(/No parse at (\d+)/);
        let position = 0;
        if (match) {
          const parsed = parseInt(match[1], 10);
          position = isNaN(parsed) ? 0 : parsed;
        }

        let inputText: string;

        if (typeof input === "string") {
          inputText = input;
        } else {
          inputText = input.read(0, input.length);
        }

        const text = Text.of(
          inputText.split("\n"),
        );

        throw createSyntacticParseError(text, position);
      }
    };
  }

  return strictSdlParser;
}
