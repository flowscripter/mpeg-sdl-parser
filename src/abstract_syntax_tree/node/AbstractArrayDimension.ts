import type Location from "../../tokenizer/token/Location.ts";
import AbstractCompositeNode from "./AbstractCompositeNode.ts";
import type ArrayDimensionKind from "./enum/array_dimension_kind.ts";
import NodeKind from "./enum/node_kind.ts";

abstract class AbstractArrayDimension extends AbstractCompositeNode {
  constructor(
    public readonly arrayDimensionKind: ArrayDimensionKind,
    location: Location,
  ) {
    super(NodeKind.ARRAY_DIMENSION, location);
  }
}

export default AbstractArrayDimension;
