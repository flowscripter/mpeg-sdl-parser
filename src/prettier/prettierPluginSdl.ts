import type { Parser, Plugin, Printer, SupportLanguage } from "prettier";
import printNode from "./print_node";
import type AbstractNode from "../ast/node/AbstractNode";
import { createStrictSdlParser } from "../lezer/createSdlParser";
import { buildAst } from "../ast/buildAst";
import SdlStringInput from "../lezer/SdlStringInput";
const languages: SupportLanguage[] = [
  {
    name: "sdl",
    parsers: ["sdl"],
  },
];

const parsers: Record<string, Parser<AbstractNode>> = {
  sdl: {
    astFormat: "sdl",
    parse: async (sdlSpecification) => {
      const sdlStringInput = new SdlStringInput(sdlSpecification);
      const sdlParser = await createStrictSdlParser();
      const parseTree = sdlParser.parse(sdlSpecification);

      return buildAst(parseTree, sdlStringInput);
    },
    locStart: (node: AbstractNode) => {
      return node.startToken.leadingTrivia.length > 0
        ? node.startToken.leadingTrivia[0].location.position
        : node.startToken.location.position;
    },
    locEnd: (node: AbstractNode) => {
      return node.endToken.trailingTrivia.length > 0
        ? node.endToken.trailingTrivia[node.endToken.trailingTrivia.length - 1]
          .location.position
        : node.endToken.location.position;
    },
  },
};

const printers: Record<string, Printer<AbstractNode>> = {
  "sdl": {
    print: printNode,
  },
};

const prettierPluginSdl: Plugin<AbstractNode> = {
  languages,
  parsers,
  printers,
};

export default prettierPluginSdl;
