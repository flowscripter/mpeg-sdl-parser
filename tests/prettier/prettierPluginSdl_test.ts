import * as prettier from "prettier";
import { promises as fs } from "node:fs";
import { type Buffer } from "node:buffer";
import path from "node:path";
import { describe, expect, test } from "bun:test";
import prettierPluginSdl from "../../src/prettier/prettierPluginSdl";
import { createStrictSdlParser } from "../../src/lezer/createSdlParser.ts";
import { SdlStringInput } from "../../src/lezer/SdlStringInput";

const strictSdlParser = createStrictSdlParser();

describe("Prettier Plugin SDL tests", () => {
  test("prettified output is as expected", async () => {
    const sampleSdlSpecification = await fs.readFile(
      path.join(__dirname, "../sample_specifications/various_elements.sdl"),
    ).then((buffer: Buffer) => buffer.toString());

    const options: prettier.Options = {
      parser: "sdl",
      plugins: [prettierPluginSdl],
    };

    const prettified = await prettier.format(sampleSdlSpecification, options);

    expect(prettified).toMatchSnapshot();
  });

  test("prettified output is valid", async () => {
    const sampleSdlSpecification = await fs.readFile(
      path.join(__dirname, "../sample_specifications/various_elements.sdl"),
    ).then((buffer: Buffer) => buffer.toString());

    const options: prettier.Options = {
      parser: "sdl",
      plugins: [prettierPluginSdl],
    };

    const prettified = await prettier.format(sampleSdlSpecification, options);

    const sdlStringInput = new SdlStringInput(prettified);
    strictSdlParser.parse(sdlStringInput);
  });
});
