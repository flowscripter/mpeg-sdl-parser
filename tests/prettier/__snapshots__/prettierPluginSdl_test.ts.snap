// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Prettier Plugin SDL tests prettified output is as expected 1`] = `
"// repeated packets in bitstream
//
// should be declared in a wrapping Bitstream class
class transport_packet {
  unsigned int(8) sync_byte = 0x47;
  bit(1) transport_error_indicator;
  bit(1) payload_unit_start_indicator;
  bit(1) transport_priority;
  unsigned int(13) // odd but valid comment position
   PID;
  bit(2)* transport_scrambling_control;
  aligned(16) bit(2) adaptation_field_control = 0x0000.0000;
  computed unsigned int N = 184;
  computed float test = 1e1;
  utf8string foo = u"bar" u"foo";
  base64string foo2 = "BAR";
  computed int dummy1 = 'ABCD' '01AB';
  computed int dummy2 = lengthof(transport_scrambling_control) + N++;
  if (adaptation_field_control == 0b10 || adaptation_field_control == 0b11) {
    adaptation_field data; // previously declared class

    // N is 184 minus the number of bytes in the adaptation_field
    N = N - // and another
    1 // and another odd comment
     - data.adaptation_field_length;
  }
  if (adaptation_field_control == 0b01 || adaptation_field_control == 0b11) {
    bit(8) data_byte[N];
  }
}

map Offsets (int) {
  0b00, { 1024 },
  0b01, { 2048 }
}

expandable(20) class odd(int a, B b) extends odder(a) : bit(2) id = 2 {
  reserved unsigned int(Offsets) offset;
  computed int myArray[2];
  computed int i = 1;
  Strange s(1, 2);
  while (i < 2) {
    i++;
  }
  i = 2;
  do {
    i--;
  } while (i > 0);
  if (1) {
    i++;
  }
  else if (1) {
    i--;
  }
  else {
    i++;
  }
  if (1)
    if (2)
      i--;
    else
      i++;
  switch (i) {
    case 0:
    case 1:
      // nothing
      break;
    default:
      i++;
    // foo
  }
}

class odd2(int a) {
  computed int x = 0;
  for (computed int j = 0; j < 2; j++) {
    x++;
  }
  for (; j > 0; j--) {
  }
} // until EOF
"
`;
